Seems like it's looking in a hardcoded place. 

```
/nix/store/51bin2x0i1v9jff0sd98b51hmzv00g0v-cctools-binutils-darwin/bin/ar cru libvex-amd64-darwin.a priv/libvex_amd64_darwin_a-main_globals.o priv/libvex_amd64_darwin_a-main_main.o priv/libvex_amd64_darwin_a-main_util.o priv/libvex_amd64_darwin_a-ir_defs.o priv/libvex_amd64_darwin_a-ir_inject.o priv/libvex_amd64_darwin_a-ir_match.o priv/libvex_amd64_darwin_a-ir_opt.o priv/libvex_amd64_darwin_a-guest_generic_bb_to_IR.o priv/libvex_amd64_darwin_a-guest_generic_x87.o priv/libvex_amd64_darwin_a-guest_x86_helpers.o priv/libvex_amd64_darwin_a-guest_x86_toIR.o priv/libvex_amd64_darwin_a-guest_amd64_helpers.o priv/libvex_amd64_darwin_a-guest_amd64_toIR.o priv/libvex_amd64_darwin_a-guest_ppc_helpers.o priv/libvex_amd64_darwin_a-guest_ppc_toIR.o priv/libvex_amd64_darwin_a-guest_arm_helpers.o priv/libvex_amd64_darwin_a-guest_arm_toIR.o priv/libvex_amd64_darwin_a-guest_arm64_helpers.o priv/libvex_amd64_darwin_a-guest_arm64_toIR.o priv/libvex_amd64_darwin_a-guest_s390_helpers.o priv/libvex_amd64_darwin_a-guest_s390_toIR.o priv/libvex_amd64_darwin_a-guest_mips_helpers.o priv/libvex_amd64_darwin_a-guest_mips_toIR.o priv/libvex_amd64_darwin_a-guest_tilegx_helpers.o priv/libvex_amd64_darwin_a-guest_tilegx_toIR.o priv/libvex_amd64_darwin_a-host_generic_regs.o priv/libvex_amd64_darwin_a-host_generic_simd64.o priv/libvex_amd64_darwin_a-host_generic_simd128.o priv/libvex_amd64_darwin_a-host_generic_simd256.o priv/libvex_amd64_darwin_a-host_generic_maddf.o priv/libvex_amd64_darwin_a-host_generic_reg_alloc2.o priv/libvex_amd64_darwin_a-host_x86_defs.o priv/libvex_amd64_darwin_a-host_x86_isel.o priv/libvex_amd64_darwin_a-host_amd64_defs.o priv/libvex_amd64_darwin_a-host_amd64_isel.o priv/libvex_amd64_darwin_a-host_ppc_defs.o priv/libvex_amd64_darwin_a-host_ppc_isel.o priv/libvex_amd64_darwin_a-host_arm_defs.o priv/libvex_amd64_darwin_a-host_arm_isel.o priv/libvex_amd64_darwin_a-host_arm64_defs.o priv/libvex_amd64_darwin_a-host_arm64_isel.o priv/libvex_amd64_darwin_a-host_s390_defs.o priv/libvex_amd64_darwin_a-host_s390_isel.o priv/libvex_amd64_darwin_a-s390_disasm.o priv/libvex_amd64_darwin_a-host_mips_defs.o priv/libvex_amd64_darwin_a-host_mips_isel.o priv/libvex_amd64_darwin_a-host_tilegx_defs.o priv/libvex_amd64_darwin_a-host_tilegx_isel.o priv/libvex_amd64_darwin_a-tilegx_disasm.o 
ranlib libvex-amd64-darwin.a
make[3]: Leaving directory '/private/var/folders/qv/5k4lhv3n1hg2jk29cj3smtp00000gn/T/nix-build-valgrind-3.12.0.drv-0/valgrind-3.12.0/VEX'
make[2]: Leaving directory '/private/var/folders/qv/5k4lhv3n1hg2jk29cj3smtp00000gn/T/nix-build-valgrind-3.12.0.drv-0/valgrind-3.12.0/VEX'
Making all in coregrind
make[2]: Entering directory '/private/var/folders/qv/5k4lhv3n1hg2jk29cj3smtp00000gn/T/nix-build-valgrind-3.12.0.drv-0/valgrind-3.12.0/coregrind'
make[2]: *** No rule to make target '/usr/include/mach/mach_vm.defs', needed by 'm_mach/mach_vmUser.c'.  Stop.
make[2]: Leaving directory '/private/var/folders/qv/5k4lhv3n1hg2jk29cj3smtp00000gn/T/nix-build-valgrind-3.12.0.drv-0/valgrind-3.12.0/coregrind'
make[1]: *** [Makefile:829: all-recursive] Error 1
make[1]: Leaving directory '/private/var/folders/qv/5k4lhv3n1hg2jk29cj3smtp00000gn/T/nix-build-valgrind-3.12.0.drv-0/valgrind-3.12.0'
make: *** [Makefile:698: all] Error 2
builder for ‘/nix/store/8npbhswdy5lkls7lz319v8w5ai42dlrd-valgrind-3.12.0.drv’ failed with exit code 2
```

I'm just running a hydra instance locally it's not related to the official one from nixos.org

--- a/pkgs/development/tools/analysis/valgrind/default.nix
+++ b/pkgs/development/tools/analysis/valgrind/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, fetchpatch, perl, gdb }:
+{ stdenv, fetchurl, fetchpatch, perl, gdb, llvm, cctools, xnu, bootstrap_cmds }:
 
 stdenv.mkDerivation rec {
   name = "valgrind-3.12.0";
@@ -14,11 +14,13 @@ stdenv.mkDerivation rec {
 
   # Perl is needed for `cg_annotate'.
   # GDB is needed to provide a sane default for `--db-command'.
-  buildInputs = [ perl ] ++ stdenv.lib.optional (!stdenv.isDarwin) gdb;
+  buildInputs = [ perl gdb ]  ++ stdenv.lib.optionals (stdenv.isDarwin) [ bootstrap_cmds xnu ];
 
   enableParallelBuilding = true;
 
-  postPatch =
+  patches = stdenv.lib.optionals (stdenv.isDarwin) [ ./valgrind-bzero.patch ];
+
+  postPatch = stdenv.lib.optionalString (stdenv.isDarwin)
     # Apple's GCC doesn't recognize `-arch' (as of version 4.2.1, build 5666).
     ''
       echo "getting rid of the \`-arch' GCC option..."
@@ -27,6 +29,23 @@ stdenv.mkDerivation rec {
 
       sed -i coregrind/link_tool_exe_darwin.in \
           -e 's/^my \$archstr = .*/my $archstr = "x86_64";/g'
+
+      echo "substitute hardcoded /usr/include/mach with ${xnu}/include/mach"
+      substituteInPlace coregrind/Makefile.in \
+         --replace /usr/include/mach ${xnu}/include/mach
+
+      echo "substitute hardcoded dsymutil with ${llvm}/bin/llvm-dsymutil"
+      find -name "Makefile.in" | while read file; do
+         substituteInPlace "$file" \
+           --replace dsymutil ${llvm}/bin/llvm-dsymutil
+      done
+
+      substituteInPlace coregrind/m_debuginfo/readmacho.c \
+         --replace /usr/bin/dsymutil ${llvm}/bin/llvm-dsymutil
+
+      echo "substitute hardcoded /usr/bin/ld with ${cctools}/bin/ld"
+      substituteInPlace coregrind/link_tool_exe_darwin.in \
+        --replace /usr/bin/ld ${cctools}/bin/ld
     '';
 
   configureFlags =
@@ -58,6 +77,6 @@ stdenv.mkDerivation rec {
     license = stdenv.lib.licenses.gpl2Plus;
 
     maintainers = [ stdenv.lib.maintainers.eelco ];
-    platforms = stdenv.lib.platforms.linux;
+    platforms = stdenv.lib.platforms.unix;
   };
 }

--- a/pkgs/development/tools/analysis/valgrind/valgrind-bzero.patch
+++ b/pkgs/development/tools/analysis/valgrind/valgrind-bzero.patch
@@ -1,0 +1,37 @@
+Index: coregrind/m_main.c
+===================================================================
+--- a/coregrind/m_main.c (revision 16102)
++++ b/coregrind/m_main.c (revision 16103)
+@@ -3489,6 +3489,10 @@
+     // skip check
+   return VG_(memset)(s,c,n);
+ }
++void __bzero(void* s, UWord n);
++void __bzero(void* s, UWord n) {
++    (void)VG_(memset)(s,0,n);
++}
+ void bzero(void *s, SizeT n);
+ void bzero(void *s, SizeT n) {
+     VG_(memset)(s,0,n);
+@@ -4058,20 +4062,7 @@
+ 
+ #endif
+ 
+-#if defined(VGO_darwin) && DARWIN_VERS == DARWIN_10_10
+ 
+-/* This might also be needed for > DARWIN_10_10, but I have no way
+-   to test for that.  Hence '==' rather than '>=' in the version
+-   test above. */
+-void __bzero ( void* s, UWord n );
+-void __bzero ( void* s, UWord n )
+-{
+-   (void) VG_(memset)( s, 0, n );
+-}
+-
+-#endif
+-
+-
+ /*--------------------------------------------------------------------*/
+ /*--- end                                                          ---*/
+ /*--------------------------------------------------------------------*/
+

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -6862,7 +6862,10 @@ with pkgs;
 
   gdb-multitarget = lowPrio (gdb.override { multitarget = true; });
 
-  valgrind = callPackage ../development/tools/analysis/valgrind { };
+  valgrind = callPackage ../development/tools/analysis/valgrind {
+    inherit (darwin) xnu bootstrap_cmds cctools;
+    llvm = llvm_39;
+   };
 
   valkyrie = callPackage ../development/tools/analysis/valkyrie { };
 

